mlrun/serving/merger.py

```diff
@@ -0,0 +1,154 @@
+
+    async def _do(self, event):
+        if event is storey.dtypes._termination_obj:
+            return await self._do_downstream(storey.dtypes._termination_obj)
+        else:
+            element = self._get_event_or_body(event)
+            fn_result = self._merge_events(element)
+            if fn_result:
+                mapped_event = self._user_fn_output_to_event(event, fn_result)
+                await self._do_downstream(mapped_event)
+
+    def _merge_events(self, event):
+        # skip if only one uplink
+        if self._uplinks <= 1:
+            return event
+
+        merge_key = self.get_join_key(event)
+        if merge_key in self._cache:
+            # old events with that key already exist (cached)
+            entry: CacheEntry = self._cache[merge_key]
+            if entry.is_late:
+                self.context.logger.info(f"event id {merge_key} arrived late")
+                return None
+
+            if self.context.verbose:
+                self.context.logger.info(
+                    f"event id {merge_key} part {entry.arrived} arrived"
+                )
+            entry.arrived += 1
+            entry.events.append(event)
+
+            if entry.arrived == self._uplinks:
+                # expected number of events arrived, can merge them
+                if self.context.verbose:
+                    self.context.logger.info(
+                        f"event {merge_key}, all {entry.arrived} parts arrived"
+                    )
+                del self._cache[merge_key]
+                self._keys_queue[entry.sequence % self._queue_len] = None
+                return self.merge_function(event, entry.events)
+        else:
+            # first time the key arrives
+            self._cache[merge_key] = CacheEntry(self._sequence, event)
+            queue_index = self._sequence % self._queue_len
+            old_key = self._keys_queue[queue_index]
+
+            if self.context.verbose:
+                self.context.logger.info(f"new event id {merge_key} arrived")
+
+            if old_key is not None:
+                # reached max queued entries, need to drop the old event
+                message = (
+                    f"missing parts for event key {old_key} after a long wait, dropping"
+                )
+                self.context.logger.warning(message)
+                if self._full_event:
+                    self.context.push_error(
+                        self._cache[old_key].events[0], f"{message}", source=self.name
+                    )
+                self._cache[old_key].is_late = True
+
+            self._keys_queue[queue_index] = merge_key
+            self._sequence += 1
+            return None
+
+    def merge_function(self, last_event, events):
+        """logic to merge all gathered events to a result event, can be overwritten by sub class
+
+        :param last_event: latest event
+        :param events: all gathered events for the same key (including latest)
+        :return: event
```

Comment:
When not using full event this returns `list[event]`

Reply:
it always returns the event/body in the default non full case the event is a list, users can define what they want to be the result event